### PR TITLE
Added Celebration Scaleform

### PR DESCRIPTION
**Pull request by Sander0542**

### CODE REVIEW

File: LemonUI/Scaleform/BaseScaleform.cs [hunk 1]
```diff
@@ -56,19 +56,32 @@ public abstract class BaseScaleform : IScaleform
...
+        }
+
+        #endregion
+
+        #region Protected Functions
+
+        /// <summary>
+        /// Load a given Scaleform
+        /// </summary>
+        protected void LoadScaleform()
```
Comment (justalemon) [inline LemonUI/Scaleform/BaseScaleform.cs:70]: Ill recommend sending this changes in a separate PR.

File: LemonUI/Scaleform/BaseScaleform.cs [hunk 2]
```diff
@@ -142,6 +155,20 @@ private void CallFunctionBase(string function, params object[] parameters)
 
         #region Public Functions
 
+        /// <summary>
+        /// Check if the Scaleform has been loaded.
+        /// </summary>
+        public bool IsLoaded()
```
Comment (justalemon) [inline LemonUI/Scaleform/BaseScaleform.cs:161]: This should also be in another PR.

Also, it should be in a Property instead of a Function.

File: LemonUI/Scaleform/Celebration.cs [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+using GTA.Native;
+#endif
+
+namespace LemonUI.Scaleform
+{
+    /// <summary>
+    /// Celebration screen that is shown at the end of a mission or job.
+    /// </summary>
+    /// <footer>
+    /// <a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration">`Celebration` on google.com</a>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:17]: Why the Google link? I feel that it serves no purpose.

File: LemonUI/Scaleform/Celebration.cs [hunk 4]
```diff
@@ -0,0 +1,502 @@
...
+    /// <a href="https://vespura.com/fivem/scaleform/#MP_CELEBRATION">`MP_CELEBRATION` on vespura.com</a>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:18]: I don't recommend linking to the Vespura site as the page has been outdated for a couple of months.

File: LemonUI/Scaleform/Celebration.cs [hunk 5]
```diff
@@ -0,0 +1,502 @@
...
+        private readonly CelebrationPart _celeb;
+        private readonly CelebrationPart _celebBg;
+        private readonly CelebrationPart _celebFg;
+
+        private readonly IList<Action<CelebrationPart>> _items;
+
+        /// <summary>
+        /// The duration of each item on the wall.
+        /// </summary>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.Duration">`Celebration.Duration` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:31]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 6]
```diff
@@ -0,0 +1,502 @@
...
+                _celeb.Duration = value;
+                _celebBg.Duration = value;
+                _celebFg.Duration = value;
+            }
+        }
+
+        /// <summary>
+        /// The list with all items.
+        /// </summary>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.Items">`Celebration.Items` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:46]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 7]
```diff
@@ -0,0 +1,502 @@
...
+                _celeb.Visible = value;
+                _celebBg.Visible = value;
+                _celebFg.Visible = value;
+            }
+        }
+
+        /// <summary>
+        /// Creates a standard Celebration with a duration of 3
+        /// </summary>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration">`Celebration` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:67]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 8]
```diff
@@ -0,0 +1,502 @@
...
+        public Celebration() : this(3)
+        {
+        }
+
+        /// <summary>
+        /// Creates a custom Celebration with the specified duration.
+        /// </summary>
+        /// <param name="duration">The duration of each item on the wall.</param>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration">`Celebration` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:76]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 9]
```diff
@@ -0,0 +1,502 @@
...
+            _celebBg = new CelebrationPart(CelebrationLayer.Background, _items);
+            _celebFg = new CelebrationPart(CelebrationLayer.Foreground, _items);
+
+            Duration = duration;
+        }
+
+        /// <summary>
+        /// Processes the object.
+        /// </summary>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.Process">`Celebration.Process` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:91]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 10]
```diff
@@ -0,0 +1,502 @@
...
+
+            _celebBg?.Process();
+            _celeb?.Process();
+            _celebFg?.Process();
+        }
+
+        /// <summary>
+        /// Remove all the items on the wall.
+        /// </summary>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.Reset">`Celebration.Reset` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:104]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 11]
```diff
@@ -0,0 +1,502 @@
...
+        /// <param name="hasWon">Whether the player has won.</param>
+        /// <param name="gamerName">The element that has been won (Game, Round, Match, etc).</param>
+        /// <param name="teamName">The team that has won the game.</param>
+        /// <param name="crewName">The crew that has won the game.</param>
+        /// <param name="betAmount">The amount that the player has gained/lost in betting (0 = off).</param>
+        /// <param name="rawGamerName">Whether the gamerName is rawText (null = automatic).</param>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.AddWinner">`Celebration.AddWinner` on google.com</a></footer>
+        public void AddWinner(bool hasWon, string gamerName = "CELEB_ROUND", string teamName = "", string crewName = "", int betAmount = 0, bool? rawGamerName = null)
+        {
+            var winLose = hasWon ? "CELEB_WINNER" : "CELEB_LOSER";
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:122]: winLose should be string instead of var. Also this can probably be completely removed as is only needed once.

File: LemonUI/Scaleform/Celebration.cs [hunk 12]
```diff
@@ -0,0 +1,502 @@
...
+        /// <summary>
+        /// Add a Winner item to the wall.
+        /// </summary>
+        /// <param name="hasWon">Whether the player has won.</param>
+        /// <param name="gamerName">The element that has been won (Game, Round, Match, etc).</param>
+        /// <param name="teamName">The team that has won the game.</param>
+        /// <param name="crewName">The crew that has won the game.</param>
+        /// <param name="betAmount">The amount that the player has gained/lost in betting (0 = off).</param>
+        /// <param name="rawGamerName">Whether the gamerName is rawText (null = automatic).</param>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.AddWinner">`Celebration.AddWinner` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:119]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 13]
```diff
@@ -0,0 +1,502 @@
...
+        /// <summary>
+        /// Add a Winner item to the wall.
+        /// </summary>
+        /// <param name="winLose">The big text that displays a message.</param>
+        /// <param name="gamerName">The element that has been won (Game, Round, Match, etc).</param>
+        /// <param name="teamName">The team that has won the game.</param>
+        /// <param name="crewName">The crew that has won the game.</param>
+        /// <param name="betAmount">The amount that the player has gained/lost in betting (0 = off).</param>
+        /// <param name="rawGamerName">Whether the gamerName is rawText (null = automatic).</param>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.AddWinner">`Celebration.AddWinner` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:136]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 14]
```diff
@@ -0,0 +1,502 @@
...
+            _items.Add(celeb => celeb.CallFunction("ADD_WINNER_TO_WALL", celeb.WallId, winLose, gamerName, crewName, betAmount, true, teamName, !rawGamerName.Value));
+        }
+
+        /// <summary>
+        /// Add a position item to the wall.
+        /// </summary>
+        /// <param name="position">The position the player is (0 = dnf).</param>
+        /// <param name="label">The label above the position (empty = off)</param>
+        /// <param name="rawLabel">Whether the label is rawText (null = automatic).</param>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.AddPosition">`Celebration.AddPosition` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:150]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 15]
```diff
@@ -0,0 +1,502 @@
...
+
+            _items.Add(celeb => celeb.CallFunction("ADD_POSITION_TO_WALL", celeb.WallId, positionStr, label, rawLabel.Value, false));
+        }
+
+        /// <summary>
+        /// Add a score item to the wall.
+        /// </summary>
+        /// <param name="value">The label with the type of score.</param>
+        /// <param name="title">The value the player has scored.</param>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.AddScore">`Celebration.AddScore` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:164]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 16]
```diff
@@ -0,0 +1,502 @@
...
+        /// </summary>
+        /// <param name="pointsGained">The amount of points the player gained.</param>
+        /// <param name="startPoints">The amount of points the player has a the start.</param>
+        /// <param name="rankMinPoints">The amount of points the rank starts at.</param>
+        /// <param name="rankMaxPoints">The amount of points needed to rank up.</param>
+        /// <param name="currentRank">The the current rank of the player.</param>
+        /// <param name="nextRank">The rank after the current rank.</param>
+        /// <param name="currentRankName">The name of the current rank (empty = off).</param>
+        /// <param name="nextRankName">The nam of the rank after the current rank (empty = off).</param>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.AddRepPointsAndRankBar">`Celebration.AddRepPointsAndRankBar` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:181]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 17]
```diff
@@ -0,0 +1,502 @@
...
+        /// </summary>
+        /// <param name="pointsGained">The amount of points the player gained.</param>
+        /// <param name="startPoints">The amount of points the player has a the start.</param>
+        /// <param name="rankMinPoints">The amount of points the rank starts at.</param>
+        /// <param name="rankMaxPoints">The amount of points needed to rank up.</param>
+        /// <param name="currentRank">The the current rank of the player.</param>
+        /// <param name="nextRank">The rank after the current rank.</param>
+        /// <param name="currentRankName">The name of the current rank (empty = off).</param>
+        /// <param name="nextRankName">The nam of the rank after the current rank (empty = off).</param>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.AddArenaPointsAndRankBar">`Celebration.AddArenaPointsAndRankBar` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:198]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 18]
```diff
@@ -0,0 +1,502 @@
...
+        public void AddArenaPointsAndRankBar(int pointsGained, int startPoints, int rankMinPoints, int rankMaxPoints, int currentRank, int nextRank, string currentRankName = "", string nextRankName = "")
+        {
+            _items.Add(celeb => celeb.CallFunction("ADD_ARENA_POINTS_AND_RANK_BAR_TO_WALL", celeb.WallId, pointsGained, startPoints, rankMinPoints, rankMaxPoints, currentRank, nextRank, currentRankName, nextRankName));
+        }
+
+        /// <summary>
+        /// Add a job points item to the wall.
+        /// </summary>
+        /// <param name="points">The amount of points.</param>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.AddJobPoint">`Celebration.AddJobPoint` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:208]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 19]
```diff
@@ -0,0 +1,502 @@
...
+        public void AddJobPoint(int points)
+        {
+            _items.Add(celeb => celeb.CallFunction("ADD_JOB_POINTS_TO_WALL", celeb.WallId, points, 0));
+        }
+
+        /// <summary>
+        /// Add a arena points item to the wall.
+        /// </summary>
+        /// <param name="points">The amount of arena points.</param>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.AddArenaPoint">`Celebration.AddArenaPoint` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:218]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 20]
```diff
@@ -0,0 +1,502 @@
...
+
+        /// <summary>
+        /// Add a mission result item to the wall.
+        /// </summary>
+        /// <param name="passed">Whether the player has passed the mission.</param>
+        /// <param name="mission">The name of the mission.</param>
+        /// <param name="missionReason">The reason why the mission has ended.</param>
+        /// <param name="rawMission">Whether the mission name is rawText (null = automatic).</param>
+        /// <param name="rawMissionReason">Whether the mission reason is rawText (null = automatic).</param>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.AddMissionResult">`Celebration.AddMissionResult` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:232]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 21]
```diff
@@ -0,0 +1,502 @@
...
+        /// <summary>
+        /// Add a mission result item to the wall.
+        /// </summary>
+        /// <param name="mission">The name of the mission.</param>
+        /// <param name="passFail">The pass or fail test.</param>
+        /// <param name="missionReason">The reason why the mission has ended.</param>
+        /// <param name="rawMission">Whether the mission name is rawText (null = automatic).</param>
+        /// <param name="rawPassFail">Whether the pass fail message is rawText (null = automatic).</param>
+        /// <param name="rawMissionReason">Whether the mission reason is rawText (null = automatic).</param>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.AddMissionResult">`Celebration.AddMissionResult` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:249]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 22]
```diff
@@ -0,0 +1,502 @@
...
+            _items.Add(celeb => celeb.CallFunction("ADD_MISSION_RESULT_TO_WALL", celeb.WallId, mission, passFail, missionReason, rawMissionReason.Value, rawPassFail.Value, rawMission.Value));
+        }
+
+        /// <summary>
+        /// Add a time item to the wall.
+        /// </summary>
+        /// <param name="time">The time.</param>
+        /// <param name="timeDiff">The difference with the time (0 = off).</param>
+        /// <param name="label">The label of the item.</param>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.AddTime">`Celebration.AddTime` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:265]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 23]
```diff
@@ -0,0 +1,502 @@
...
+        public void AddTime(int time, int timeDiff = 0, string label = "CELEB_TIME")
+        {
+            _items.Add(celeb => celeb.CallFunction("ADD_TIME_TO_WALL", celeb.WallId, time, label, timeDiff));
+        }
+
+        /// <summary>
+        /// Add a cash item to the wall.
+        /// </summary>
+        /// <param name="cash">The amount of cash.</param>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.AddCash">`Celebration.AddCash` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:275]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 24]
```diff
@@ -0,0 +1,502 @@
...
+        public void AddCash(int cash)
+        {
+            _items.Add(celeb => celeb.CallFunction("ADD_CASH_TO_WALL", celeb.WallId, cash, 0));
+        }
+
+        /// <summary>
+        /// Add a world record item to the wall.
+        /// </summary>
+        /// <param name="time">The time of the record.</param>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.AddWorldRecord">`Celebration.AddWorldRecord` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:285]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 25]
```diff
@@ -0,0 +1,502 @@
...
+
+        /// <summary>
+        /// Add a tournament item to the wall.
+        /// </summary>
+        /// <param name="playlistName">The name of the playlist.</param>
+        /// <param name="resultText">The name of the tournament.</param>
+        /// <param name="resultValue">The value of the tournament.</param>
+        /// <param name="qualification">The label with the message.</param>
+        /// <param name="rawResult">Whether the result is rawText (null = automatic).</param>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.AddTournament">`Celebration.AddTournament` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:299]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 26]
```diff
@@ -0,0 +1,502 @@
...
+            _items.Add(celeb => celeb.CallFunction("ADD_TOURNAMENT_TO_WALL", celeb.WallId, playlistName, qualification, resultText, rawResult.Value, resultValue));
+        }
+
+        /// <summary>
+        /// Add a objective item to the wall.
+        /// </summary>
+        /// <param name="title">The title of the objective.</param>
+        /// <param name="text">The text of the objective.</param>
+        /// <param name="rawTitle">Whether the title is rawText (null = automatic).</param>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.AddObjective">`Celebration.AddObjective` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:313]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 27]
```diff
@@ -0,0 +1,502 @@
...
+            _items.Add(celeb => celeb.CallFunction("ADD_OBJECTIVE_TO_WALL", celeb.WallId, title, text, rawTitle.Value));
+        }
+
+        /// <summary>
+        /// Add a numeric stat item to the wall.
+        /// </summary>
+        /// <param name="stat">The name of the stat.</param>
+        /// <param name="value">The value of the stat</param>
+        /// <param name="rawStat">Whether the stat is rawText (null = automatic).</param>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.AddStatNumeric">`Celebration.AddStatNumeric` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:327]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 28]
```diff
@@ -0,0 +1,502 @@
...
+
+            _items.Add(celeb => celeb.CallFunction("ADD_STAT_NUMERIC_TO_WALL", celeb.WallId, stat, value, 0, rawStat.Value));
+        }
+
+        /// <summary>
+        /// Add a wave reached item to the wall.
+        /// </summary>
+        /// <param name="value">The wave that has been reached.</param>
+        /// <param name="title">The label with the title that has been reached.</param>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.AddWaveReached">`Celebration.AddWaveReached` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:340]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 29]
```diff
@@ -0,0 +1,502 @@
...
+        public void AddWaveReached(int value, string title = "BM_WAVE_COMP")
+        {
+            _items.Add(celeb => celeb.CallFunction("ADD_WAVE_REACHED_TO_WALL", celeb.WallId, value, title));
+        }
+
+        // public void AddIntro(string jobName, string challenge, string challengeParts, string targetType, string targetValue, int delay, string targetValuePrefix, string mode = Mode.Mission, bool? rawMode = null) //TODO (Not working yet)
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:346]: If this is not being used, it should be removed.

File: LemonUI/Scaleform/Celebration.cs [hunk 30]
```diff
@@ -0,0 +1,502 @@
...
+        //     if (!rawMode.HasValue) rawMode = IsRawText(mode);
+        //
+        //     _items.Add(celeb => celeb.CallFunction("ADD_INTRO_TO_WALL", celeb.WallId, mode, jobName, challenge, challengeParts, targetType, targetValue, delay, targetValuePrefix, true, "HUD_COLOUR_GREEN"));
+        // }
+
+        /// <summary>
+        /// Add a post unlock cash item to the wall.
+        /// </summary>
+        /// <param name="cash">The amount of cash</param>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.AddPostUnlockCash">`Celebration.AddPostUnlockCash` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:357]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 31]
```diff
@@ -0,0 +1,502 @@
...
+        /// <summary>
+        /// Add a challenge winner item to the wall.
+        /// </summary>
+        /// <param name="challenge">The name of the challenge.</param>
+        /// <param name="winner">The winner of the challenge.</param>
+        /// <param name="isMission">Whether the challenge is a mission.</param>
+        /// <param name="hasWon">Whether the player has won the challenge.</param>
+        /// <param name="crew">The name of the crew that has won the challenge.</param>
+        /// <param name="bet">The bet amount the player a gained/lost (0 = off).</param>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.AddChallengeWinner">`Celebration.AddChallengeWinner` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:372]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 32]
```diff
@@ -0,0 +1,502 @@
...
+        /// Add a challenge winner item to the wall.
+        /// </summary>
+        /// <param name="challenge">The name of the challenge.</param>
+        /// <param name="winner">The winner of the challenge.</param>
+        /// <param name="isMission">Whether the challenge is a mission.</param>
+        /// <param name="winLose">The winLose text.</param>
+        /// <param name="crew">The name of the crew that has won the challenge.</param>
+        /// <param name="bet">The bet amount the player a gained/lost (0 = off).</param>
+        /// <param name="rawWinLose">Whether the win lose text is rawText (null = automatic).</param>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.AddChallengeWinner">`Celebration.AddChallengeWinner` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:390]: Same here with the Google links.

File: LemonUI/Scaleform/Celebration.cs [hunk 33]
```diff
@@ -0,0 +1,502 @@
...
+
+            _items.Add(celeb => celeb.CallFunction("ADD_CHALLENGE_WINNER_TO_WALL", celeb.WallId, challenge, winLose, crew, winner, bet, true, true, isMission, rawWinLose.Value, ""));
+        }
+
+        /// <summary>
+        /// Check whether text is rawText or a label.
+        /// </summary>
+        /// <param name="text">The text to be checked.</param>
+        /// <returns>Whether the text is rawText.</returns>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.Celebration.IsRawText">`Celebration.IsRawText` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/Celebration.cs:403]: Same here with the Google links.

File: LemonUI/Scaleform/CelebrationPart.cs [hunk 34]
```diff
@@ -0,0 +1,166 @@
...
+#elif (SHVDN2 || SHVDN3)
+using GTA.Native;
+#endif
+
+namespace LemonUI.Scaleform
+{
+    /// <summary>
+    /// Celebration screen that is shown at the end of a mission or job.
+    /// </summary>
+    /// <a href="https://www.google.com/search?q=LemonUI.Scaleform.CelebrationPart">`CelebrationPart` on google.com</a>
```
Comment (justalemon) [inline LemonUI/Scaleform/CelebrationPart.cs:16]: Same here with the Google links.

File: LemonUI/Scaleform/CelebrationPart.cs [hunk 35]
```diff
@@ -0,0 +1,166 @@
...
+    /// <a href="https://vespura.com/fivem/scaleform/#MP_CELEBRATION">`MP_CELEBRATION` on vespura.com</a>
```
Comment (justalemon) [inline LemonUI/Scaleform/CelebrationPart.cs:17]: Same here with the Vespura links.

File: LemonUI/Scaleform/CelebrationPart.cs [hunk 36]
```diff
@@ -0,0 +1,166 @@
...
+        private int _durationValue = 0;
+        private int _duration = -1;
+        private int _start;
+
+        private readonly IList<Action<CelebrationPart>> _items;
+
+        /// <summary>
+        /// The duration of each item on the wall.
+        /// </summary>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.CelebrationPart.Duration">`CelebrationPart.Duration` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/CelebrationPart.cs:33]: Same here with the Google links.

File: LemonUI/Scaleform/CelebrationPart.cs [hunk 37]
```diff
@@ -0,0 +1,166 @@
...
+        public int Duration { get; set; }
+
+        /// <summary>
+        /// If the Scaleform should be visible or not.
+        /// </summary>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.CelebrationPart.Visible">`CelebrationPart.Visible` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/CelebrationPart.cs:39]: Same here with the Google links.

File: LemonUI/Scaleform/CelebrationPart.cs [hunk 38]
```diff
@@ -0,0 +1,166 @@
...
+                    Hide();
+                }
+                base.Visible = value;
+            }
+        }
+
+        /// <summary>
+        /// The Id of the wall.
+        /// </summary>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.CelebrationPart.WallId">`CelebrationPart.WallId` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/CelebrationPart.cs:56]: Same here with the Google links.

File: LemonUI/Scaleform/CelebrationPart.cs [hunk 39]
```diff
@@ -0,0 +1,166 @@
...
+        public string WallId => "ending";
+
+        /// <summary>
+        /// Creates a standard part for the Celebration
+        /// </summary>
+        /// <param name="layer">The type of layer for this part.</param>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.CelebrationPart">`CelebrationPart` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/CelebrationPart.cs:63]: Same here with the Google links.

File: LemonUI/Scaleform/CelebrationPart.cs [hunk 40]
```diff
@@ -0,0 +1,166 @@
...
+        internal CelebrationPart(CelebrationLayer layer) : this(layer, new List<Action<CelebrationPart>>())
+        {
+        }
+
+        /// <summary>
+        /// Creates a custom CelebrationPart with a given list of items.
+        /// </summary>
+        /// <param name="layer"></param>
+        /// <param name="items"></param>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.CelebrationPart">`CelebrationPart` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/CelebrationPart.cs:73]: Same here with the Google links.

File: LemonUI/Scaleform/CelebrationPart.cs [hunk 41]
```diff
@@ -0,0 +1,166 @@
...
+        internal CelebrationPart(CelebrationLayer layer, IList<Action<CelebrationPart>> items) : base(layer.Scaleform())
+        {
+            _layer = layer;
+            _items = items;
+        }
+
+        /// <summary>
+        /// Load the Scaleform and start the show the wall when it's done loading.
+        /// </summary>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.CelebrationPart.Update">`CelebrationPart.Update` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/CelebrationPart.cs:83]: Same here with the Google links.

File: LemonUI/Scaleform/CelebrationPart.cs [hunk 42]
```diff
@@ -0,0 +1,166 @@
...
+            else if (GetGameTimer() - _start > _duration)
+            {
+                Visible = false;
+            }
+        }
+
+        /// <summary>
+        /// Build the wall and make it show it on the Scaleform.
+        /// </summary>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.CelebrationPart.Show">`CelebrationPart.Show` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/CelebrationPart.cs:112]: Same here with the Google links.

File: LemonUI/Scaleform/CelebrationPart.cs [hunk 43]
```diff
@@ -0,0 +1,166 @@
...
+
+            _durationValue = CallFunctionReturn("GET_TOTAL_WALL_DURATION");
+            CallFunction("SHOW_STAT_WALL", WallId);
+            _start = GetGameTimer();
+        }
+
+        /// <summary>
+        /// Cleanup the wall and reset the scaleform for the next use.
+        /// </summary>
+        /// <footer><a href="https://www.google.com/search?q=LemonUI.Scaleform.CelebrationPart.Hide">`CelebrationPart.Hide` on google.com</a></footer>
```
Comment (justalemon) [inline LemonUI/Scaleform/CelebrationPart.cs:133]: Same here with the Google links.

File: LemonUI/Scaleform/CelebrationPart.cs [hunk 44]
```diff
@@ -0,0 +1,166 @@
...
+            NativeFunction.CallByHash<int>(0x1D132D614DD86811, new NativeArgument(id));
+#elif (SHVDN2 || SHVDN3)
+            Function.Call(Hash.SET_SCALEFORM_MOVIE_AS_NO_LONGER_NEEDED, new OutputArgument(id));
+#endif
+        }
+
+        /// <summary>
+        /// Get the current game timer.
+        /// </summary>
+        private int GetGameTimer()
```
Comment (justalemon) [inline LemonUI/Scaleform/CelebrationPart.cs:155]: There is Game.GameTimer that does this exact same thing.

**Review by justalemon — CHANGES_REQUESTED** [review]
Looks promising! But there are a couple of things that I feel should be fixed before merging.

File: LemonUI/Scaleform/Celebration.cs (hunk 3 above)
Comment (Sander0542) [reply LemonUI/Scaleform/Celebration.cs:17]: I added the Google links because it was done in some other classes too. I'll remove them

File: LemonUI/Scaleform/BaseScaleform.cs (hunk 1 above)
Comment (Sander0542) [reply LemonUI/Scaleform/BaseScaleform.cs:70]: I'll make a new PR with only those changes

File: LemonUI/Scaleform/Celebration.cs (hunk 4 above)
Comment (Sander0542) [reply LemonUI/Scaleform/Celebration.cs:18]: I'll remove this link too

**Comment by Sander0542** [comment]
I fixed all the request and created another PR for the BaseScaleform loading (#46). This PR also contains those commits, so that PR needs to be merged first.

**Review by justalemon — CHANGES_REQUESTED** [review]

**Comment by justalemon** [comment]
Hey sorry for not checking this before, I was on vacation.

If you still want me to check and merge this PR, feel free to reopen it :)

**Comment by Sander0542** [comment]
I found some errors in the code which made the scaleforms not load or load at different times. When I fix the issues out I'll open this PR again.
